src/app/core/services/app-server/app-server.service.ts

```diff
@@ -0,0 +1,172 @@
+  initApiClient() {
+    return Promise.all([this.getAppServerURL(), this.token.getTokens()]).then(
+      ([url, tokens]) =>
+        Swagger({
+          url: url + this.API_DOCS_URL,
+          requestInterceptor: req => {
+            req.headers['Authorization'] = 'Bearer ' + tokens.access_token
+            req.headers['Content-Type'] = DefaultRequestJSONContentType
+          }
+        })
+          .then(client => (this.apiClient = client))
+          .catch(e => this.logger.error('Error pulling API docs', e))
+    )
+  }
+
+  getApiClient() {
+    if (this.apiClient) return Promise.resolve(this.apiClient)
+    else return this.initApiClient().then(() => this.apiClient)
+  }
+
+  checkProjectAndSubjectExistElseCreate(): Promise<any> {
+    return this.checkProjectExistsElseCreate().then(() =>
+      this.checkSubjectExistsElseCreate()
+    )
+  }
+
+  checkProjectExistsElseCreate(): Promise<any> {
+    // NOTE: Adding retries here because of random 'Failed to load resource'
+    let attempts = 0
+    return this.subjectConfig.getProjectName().then(projectId => {
+      return this.getApiClient().then(apiClient =>
+        apiClient.apis[this.RADAR_PROJECT_CONTROLLER]
+          .getProjectsUsingProjectId({ projectId })
+          .catch(e => {
+            if (e.status == 404) return this.addProjectToServer(projectId)
+            else if (attempts++ < this.MAX_API_RETRIES)
+              return this.checkProjectExistsElseCreate()
+          })
+      )
+    })
+  }
+
+  addProjectToServer(projectId) {
+    return this.getApiClient().then(apiClient =>
+      apiClient.apis[this.RADAR_PROJECT_CONTROLLER].addProject(
+        {},
+        {
+          requestBody: {
+            projectId
+          }
+        }
+      )
+    )
+  }
+
+  checkSubjectExistsElseCreate(): Promise<any> {
+    return Promise.all([
+      this.subjectConfig.getParticipantLogin(),
+      this.subjectConfig.getProjectName(),
+      this.subjectConfig.getEnrolmentDate(),
+      this.getFCMToken()
+    ]).then(([subjectId, projectId, enrolmentDate, fcmToken]) => {
+      return this.getApiClient().then(apiClient =>
+        apiClient.apis[this.RADAR_USER_CONTROLLER]
+          .getRadarUserUsingSubjectId({ subjectId })
+          .then(res => res.body)
+          .catch(e => {
+            if (e.status == 404)
+              return this.addSubjectToServer(
+                subjectId,
+                projectId,
+                enrolmentDate,
+                fcmToken
+              )
+            else Promise.reject(e)
+          })
+      )
+    })
+  }
+
+  addSubjectToServer(subjectId, projectId, enrolmentDate, fcmToken) {
+    return this.getApiClient().then(apiClient =>
+      apiClient.apis[this.RADAR_USER_CONTROLLER].addUserToProject(
+        { projectId },
+        {
+          requestBody: {
+            enrolmentDate: new Date(enrolmentDate),
+            projectId,
+            subjectId,
+            fcmToken,
+            timezone: moment.tz.guess(),
```

Comment:
Should we take the stored timezone in the app? Will help keep the data consistent?

Reply:
Yeah in the app we're storing the UTC offset, but the app server needs the string (e.g. `Europe/London`), and I thought this was the easiest way to get the string. But I'll change it to refer to the stored timezone.